src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt

```diff
@@ -24,13 +48,137 @@ interface ICriterionService {
+    private val userRepo: IUserTableRepo,
+    private val projectRepo: IProjectTableRepo,
+    private val projectMemberRepo: IProjectMemberTableRepo,
 ) : ICriterionService {
+    /**
+     * Validates specific preconditions for the given criterion. A valid criterion must have
+     * either a non-null `projectId` or a non-null `createdBy`, but not both.
+     * If the precondition is violated, a [SnowballRException.FailedPreconditionException] is thrown.
+     *
+     * @param criterion The criterion object to be validated. It is expected to have a
+     *                  combination of fields that meet the required precondition.
+     * @throws SnowballRException.FailedPreconditionException If the criterion does not satisfy the required conditions.
+     */
+    private fun checkPreconditions(criterion: Criterion) {
+        val bothNull = criterion.projectId == null && criterion.createdBy == null
+        val bothNotNull = criterion.projectId != null && criterion.createdBy != null
+        if (bothNull || bothNotNull) {
+            throw SnowballRException.FailedPreconditionException("The criterion is not valid anymore.")
+        }
+    }
+
+    /**
+     * Checks if the given user has the required permissions to access or modify a project criterion.
+     *
+     * This function verifies if the user has permission to access or update a criterion
+     * that belongs to an active project and throws exceptions when access is unauthorized
+     * or when attempting to update a criterion of an inactive project.
+     *
+     * @param criterion The criterion for which permissions need to be checked. It may or may not
+     *                  be associated with a project.
+     * @param currentUser The user whose permissions are being validated.
+     * @param requestId The identifier of the request being performed, used in exception messages.
+     * @param accessType The type of access being requested (e.g., READ, UPDATE).
+     *
+     * @throws SnowballRException.UnauthorizedException If the user does not have access permissions
+     *         to the criterion.
+     * @throws SnowballRException.FailedPreconditionException If attempting to update a criterion
+     *         in a project that is not active.
+     */
+    private suspend fun checkProjectCriterionPermission(
+        criterion: Criterion,
+        currentUser: User,
+        requestId: String,
+        accessType: AccessType,
+    ) {
+        if (criterion.projectId != null && criterion.createdBy == null) {
+            val project = projectRepo.getProjectById(criterion.projectId)
+            val members = when (accessType) {
+                AccessType.READ -> projectMemberRepo.getMembersOfProject(project.id)
+                AccessType.UPDATE -> projectMemberRepo.getAllProjectAdmins(project.id)
+                else -> emptyList()
+            }
+
+            if (!members.map { member -> member.userId }.contains(currentUser.id)) {
```

Comment:
`map` and `contains` can always be simplified to an `any`, similar to the `some` method in JS.
```suggestion
            if (!members.any { it.userId == currentUser.id }) {
```

Optionally, we could store this query in a variable and then use the variable in the condition to make it more readable.

Reply:
That as good help. Thanks.